[worker] feat: Support Async reward calculation.

chrisyao19 opened the pull request:
### Checklist Before Starting

- [x] Search for similar PR(s).

### What does this PR do?

Support async reward calculation for IO bound reward functions, refactored reward managers with OOP oriented style.

### Specific Changes

1. Implemented all share logic in the base manager for all reward managers.
2. Support async reward calculation for Naive / DAPO reward managers, detect user async reward function and enable async calculation automatically.
3. Support QPS, max concurrency and timeout for async rewards.

### Test

Executed all 4 reward managers with GSM8K RL, calculated rewards with both new / existing implementations concurrently, then  `assert torch.allclose(new_reward_tensor, original_reward_tensor, rtol=1e-05, atol=1e-08)`

### Additional Info.

- **Issue Number**: Fixes issue #1210
- **Training**: None
- **Inference**: None

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/volcengine/verl?tab=readme-ov-file#contribution-guide).
- [x] Apply [pre-commit checks](https://github.com/volcengine/verl?tab=readme-ov-file#code-linting-and-formatting).
- [x] Add `[BREAKING]` to the PR title if it breaks any API.
- [x] Update the documentation about your changes in the [docs](https://github.com/volcengine/verl/tree/main/docs).
- [x] New CI unit test(s) are added to cover the code path.
- [x] Rely on existing unit tests on CI that covers the code path.

jybsuper commented on verl/workers/reward_manager/naive.py:
Could you mention the support for async reward calculation in the [config.rst](https://github.com/volcengine/verl/blob/main/docs/examples/config.rst?plain=1#L419)?

chrisyao19 replied:
Done.

jybsuper commented on verl/trainer/config/ppo_trainer.yaml:
Given all kwargs for user defined reward function are all passed to rewrad manager via `custom_reward_function.reward_kwargs`, could you add `reward_kwargs: {}` to this and megatron config?

vermouth1992 commented:
Could you help add a test?

eric-haibin-lin reviewed (commented):
i made some refactoring in the main branch, please remember to merge with main. thx!

Qin10 commented:
Hello~
I encountered the following error after merging your code into the main branch (it worked fine before the merge):
![image](https://github.com/user-attachments/assets/3085f23a-1387-4e45-8a30-2cc0da210bc9)

Qin10 commented:
Hello～There is a circular import issue in dapo.py, and it can be fixed by changing to relative imports (Naive).
![image](https://github.com/user-attachments/assets/287960c3-610e-4c69-9186-8cca6a35e19e)

chrisyao19 commented:
@vermouth1992 @eric-haibin-lin 

Hi, I have added tests and rebased on the main branch, could you help take a look? Thanks.

chrisyao19 commented:
> @vermouth1992 @eric-haibin-lin
> 
> Hi, I have added tests and rebased on the main branch, could you help take a look? Thanks.

@vermouth1992 @eric-haibin-lin 

The old base seems some issues and the CI tests failed. I have synced the most up to date commit, could you restart the CI test again? Thanks.

jybsuper commented on verl/workers/reward_manager/base.py:
```diff
@@ -0,0 +1,115 @@
+from abc import ABC, abstractmethod
```
@chrisyao19 License test fails due to missing license for this file. Could you add the following license?
```
# Copyright 2025 Individual Contributor: Lianyu Yao
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
```

chrisyao19 commented:
> @vermouth1992 @eric-haibin-lin
> 
> The old base seems some issues and the CI tests failed. I have synced the most up to date commit, could you restart the CI test again? Thanks.

@vermouth1992 @eric-haibin-lin
Hi I fixed CI test and made a rebase, could you help take another pass and trigger workflow again? Thanks and sorry for the back and forth!

chrisyao19 commented:
@vermouth1992 @eric-haibin-lin fixed the CI test(forgot to add license to the test file), please take a look when you have time, thanks!

eric-haibin-lin commented on verl/trainer/config/ppo_trainer.yaml:
```diff
@@ -765,6 +765,15 @@ reward_model:
   # the reward manager can be set to prime for parallel verification.
   reward_manager: naive
 
```
could u merge with main before we starting the CI? thx

chrisyao19 commented:
Hi @vermouth1992 @eric-haibin-lin, I've resolved the conflict on this PR, could you please take a look when you have time and let me know if you have any comments on this PR. thanks!

CatIIIIIIII commented:
this push somehow has bugs, when I use testcases to test generated response, it performs sequentially no matter how to set the concorrency.
